[DELIVERY-652] Hydrolix exporter implementation

jjohnson-hdb opened the pull request.

cap10morgan commented on src/services/exporter.ts:
```diff
@@ -0,0 +1,20 @@
+
+	abstract initSession(): Promise<void>;
+	abstract publishLogs(logs: Log[]): Promise<void>;
+	abstract publishMetrics(metrics: Metrics): Promise<void>;
```
Some mixed indentation here between line 11 and the rest

jjohnson-hdb replied:
ive noticed this happen for me from time to time in github... all fine in my ide but an indentation is off in github - not sure why but will try to fix

cap10morgan reviewed (commented):
I had some pretty high-level questions, so I stopped reviewing line by line until those are addressed.

jjohnson-hdb commented on README.md:
```diff
@@ -1,23 +1,77 @@
-# Hydrolix exporter
+# Hydrolix Logs Exporter
```
agreed will update naming

cap10morgan replied:
Sounds good. And I now understand we want both logs and analytics to be exportable by this, so I'll stop questioning that. :)

jjohnson-hdb commented on README.md:
```diff
@@ -1,23 +1,77 @@
+
+To run the component, you will need to set up a `.env` file in the root of the component directory, and a `config.json` file in `/src`. 
+
+Copy the `.env.example` file to `.env`, and `/src/config-example.json` to `/src/config.json`. Fill in appropriate values.
```
yes agreed

kriszyp commented on README.md:
```diff
@@ -1,23 +1,77 @@
+## Overview
 
-## Status
+This is a Harper component designed to export system logs and analytics to data lakes. It has been initially designed to support Hydrolix, but could easily be extended to support other platforms.
```
Yeah, "easily" extended might be stretch. And if someone did that, changing the sentence about Exporting to Hydrolix would be the easiest part to change :).

jjohnson-hdb replied:
I'll just focus specifically on hydrolix for now

kriszyp reviewed (commented):
If I understand correctly, this PR is adding support for log exporting to hydrolix exporter so that it can be used to export logs or analytics or both? I realize this is draft, so maybe you haven't had a chance to describe this work more fully.

jjohnson-hdb commented on README.md:
```diff
@@ -1,23 +1,77 @@
+| Option                | Description                                                                                          | Default |
+| --------------------- | ---------------------------------------------------------------------------------------------------- | ------- |
+| `logLevel`            | The level of logging to be used. Options: 'all', 'notify', 'error', 'warn', 'info', 'debug', 'trace' | `all`   |
+| `includeSystemInfo`   | Whether to export system analytics information. Options: true, false                                 | `true`  |
```
you could set the logIngestPercentage to 0 and it would accomplish that

cap10morgan replied:
> you could set the logIngestPercentage to 0 and it would accomplish that

I don't fully understand how that works / why it exists. Perhaps as a side effect of that, this seems unintuitive to me.

jjohnson-hdb replied:
Jeff had indicated that there should be the ability to the percentage of logs that actually get exported (because perhaps a client can't afford to or otherwise does not want to have to store 100% of the logs) so this setting is to handle that. When the value is < 1 logs will be sampled by skipping every X number of logs based on the percentage and the number of logs obtained. if 0 this would effectively disable exporting logs altogether

jjohnson-hdb replied:
i could probably include some more descriptive language about this in the readme

jjohnson-hdb commented on README.md:
```diff
@@ -1,23 +1,77 @@
+| Endpoint            | Description                                             |
+| ------------------- | ------------------------------------------------------- |
+| GET `/exportConfig/:export_destination` | REST endpoint to view export configuration properties   |
+| PUT `/exportConfig/:export_destination` | REST endpoint to update export configuration properties |
```
Yes to the update part..

The reason i didnt simply do `/config` is because I thought if this is deployed to an instance running multiple components that endpoint could be confusing due to the lack of specificity, no?

Perhaps something like this would be better? `/hydrolix/config` ?

cap10morgan replied:
Yeah, good thinking to namespace it. Maybe `/hydrolix-exporter/config`? Makes it very explicit what it is.

jjohnson-hdb commented on README.md:
```diff
@@ -1,23 +1,77 @@
+## Hydrolix Tables & Transforms
+
+To export logs to Hydrolix, you will need to [create a table](https://docs.hydrolix.io/reference/config_v1_orgs_projects_tables_create) and [corresponding transform](https://docs.hydrolix.io/reference/config_v1_orgs_projects_tables_transforms_create) in Hydrolix.
+One table/transform pair is required for logs, and one for system analytics (if enabled). See `/transformTemplates` for example JSON to create the transforms.
```
(added the auto transform creation back)

cap10morgan commented on package-lock.json:
Haha, yeah, it's a grey area. If installed via npm then it's pretty library-like and yeah this won't come into play. For now the installation instructions start with "clone this repo" but I suppose they won't stay that way for long.

cap10morgan commented on resources.ts:
```diff
@@ -1,385 +0,0 @@
-import { server, databases, logger } from 'harperdb';
```
Actual code organization FTW! ;)

I was planning to do something similar if my work on this had continued. It's hard to represent cleanly in git, though. You have to reorganize the code first, resisting the urge to make any but the necessary-to-build-and-run-it changes, and then commit that on its own. If I had a nickel for every time I've remembered to do that... I wouldn't have that many nickels.

jjohnson-hdb commented:
Summarizing recent changes:
* Refocused implementation around Hydrolix specifically as opposed to attempting to generalize
* removed use of system_information operation (i will return to this in the near future and add functionality for exporting analytics via the analytics table but i'll need to jump over to other priorities first)
* Impproved some naming convention and added some additional README detail

cap10morgan commented:
This is looking good!

Is there a convention or restriction on the resource having hyphens in its path segment (I should probably know the answer to that haha)? If not, I think `/hydrolix-exporter/config` (hyphen instead of underscore) would be better so that it matches the component's name.

jjohnson-hdb commented:
@cap10morgan 

> Is there a convention or restriction on the resource having hyphens in its path segment (I should probably know the answer to that haha)? If not, I think /hydrolix-exporter/config (hyphen instead of underscore) would be better so that it matches the component's name.

I would definitely prefer the `hydrolix-exporter` convention but i couldn't figure out how to export it that way... perhaps I'm just being silly but the resource is currently being exported like this:

```
export const hydrolix_exporter = {
	config: ExportConfig,
};
```

the issue is that we want `hydrolix-exporter` to be the root of the path, and we can't name a variable like that... if it was not the root of th path i could do something like this

```

export const path = {
	'hydrolix-exporter': {
	  config: ExportConfig,
	},
};
  
```

but that would make the resulting route `/path/hydrolix-exporter/config`

do you know of other ways to approach this?

cap10morgan commented:
@jjohnson-hdb Do the `path` options in here help? https://docs.harperdb.io/docs/developers/applications#configuring-applications-components

Perhaps with the above you could set the path to `/hydrolix-exporter` and then export a `config` resource under that?

jjohnson-hdb commented:
aha, yes! that definitely helps
